src/cargo/util/rustc.rs

```diff
@@ -38,14 +38,31 @@ impl Rustc {
     /// If successful this function returns a description of the compiler along
     /// with a list of its capabilities.
     pub fn new(
-        path: PathBuf,
+        mut path: PathBuf,
         wrapper: Option<PathBuf>,
         workspace_wrapper: Option<PathBuf>,
         rustup_rustc: &Path,
         cache_location: Option<PathBuf>,
     ) -> CargoResult<Rustc> {
         let _p = profile::start("Rustc::new");
 
+        // In order to avoid calling through rustup multiple times, we first ask
+        // rustc to give us the "resolved" rustc path, and use that instead. If
+        // this doesn't give us a path, then we just use the original path such
+        // that the following logic can handle any resulting errors normally.
+        let mut cmd = ProcessBuilder::new(&path);
```

Comment:
I would also like sign off from rustup folks on this being a viable alternative to #3035 (ie they like it enough not to continue down that path, making this obsolete)